src/server.c

```diff
@@ -1745,15 +1745,20 @@ void beforeSleep(struct aeEventLoop *eventLoop) {
      * since the unblocked clients may write data. */
     handleClientsBlockedOnKeys();
 
+    /* Record time consumption of AOF writing. */
+    long long duration_aof_start = getMonotonicUs();
     /* Record cron time in beforeSleep. This does not include the time consumed by AOF writing and IO writing below. */
-    duration_before_aof = getMonotonicUs() - duration_before_aof;
+    duration_before_aof = duration_aof_start - duration_before_aof;
 
     /* Write the AOF buffer on disk,
      * must be done before handleClientsWithPendingWritesUsingThreads,
      * in case of appendfsync=always. */
     if (server.aof_state == AOF_ON || server.aof_state == AOF_WAIT_REWRITE)
         flushAppendOnlyFile(0);
 
+    /* Record time consumption of AOF writing. */
+    durationAddSample(EL_DURATION_TYPE_AOF, getMonotonicUs() - duration_aof_start);
```

Comment:
ohh, i see what i was missing.
i was aiming to completely drop `duration_before_aof` and `duration_after_write`, and instead measure in one variable the entire duration of beforeSleep, and in another the entire duration of the AOF part and then deduct one from the other. (would not have required to add any additional time sampling call).

but i missed the fact we also need to skip counting handleClientsWithPendingWritesUsingThreads as cron.

well, looking at that, since handleClientsWithPendingWritesUsingThreads is only called from one place (here), i think we can move it's measurement from inside the function to outside of it (and avoid the complication of a separate measurement when that function early exits due to no threaded IO).
that would mean we're also counting a few other minor things as if they are IO, i.e. spinning up threads, and doing memory accounting, but i think that's acceptable.

in that case we'll have beforeSlep measure it's total time (one measurement at at the beginning and one at the end.
and we'll have an additional 3 calls to time sampling (before the AOF part, between the AOF and writes, and after writes), computing two additional variables (AOF duration and write duration) to be deducted from the total time at the end.

WDYT?

Reply:
It is a little bit asymmetric taking samples of IO reading inside of `handleClientsWithPendingReadsUsingThreads`(I don't think we should move it outside as this function also processes commands), but take samples of IO writing outside of `handleClientsWithPendingReadsUsingThreads`.

...but I guess it is OK, since I already moved the sampling of AOF into `beforeSleep`.

Reply:
ohh, right. i now notice it calls processPendingCommandAndInputBuffer, we can't count that as either IO or cron chores.
i think we need both the counting inside that function as IO, and also the counting outside it to deduct the whole thing from cron chores.
i.e. your current code.

i also notice that it processPendingCommandAndInputBuffer is called from a few other places (i missed that earlier, must have been looking at a different branch.

Reply:
took another look at beforeSleep, i see we currently include the calls in processUnblockedClients, handleClientsBlockedOnKeys in the measurement, i think we must skip them.
i don't feel comfortable re-ordering that function, but i also don't like a ton of measurements (hard to maintain).
need to think what to do here.. maybe change the plan...

Reply:
maybe it should be exclusive rather than inclusive, we do want:
* clusterBeforeSleep
* activeExpireCycle
* moduleFireServerEvent

optional:
* moduleHandleBlockedClients - can be considered a server side thing (doesn't represent commands or IO)
* evictClients - i'm assuming it's usually not very intensive (unlike IO, or something like commands that can have unexpected complexity)
* trackingBroadcastInvalidationMessages - same
* freeClientsInAsyncFreeQueue - same
* handleBlockedClientsTimeout - same

exclude (IO and commands):
* flushAppendOnlyFile
* handleClientsBlockedOnKeys
* handleClientsWithPendingReadsUsingThreads
* handleClientsWithPendingWritesUsingThreads

so either we measure everything and exclude 4 blocks (the last list above).
or we include just 3 specific blocks (the first list above).

@madolson please advise.

Reply:
Not sure about the ask. I think we should measure everything and exclude the four blocks to have an "other" category, I wouldn't call it a cron.

Reply:
ohh, now i see your other comment below.
so you just wanna keep the `total` (from beforeSleep to afterSleep), and `cmd`.
and drop all the IO and cron chores?

i think it is important to know what's slowing things down besides commands, and if it's IO or other things.
we do have LATENCY HISTORY measurement for AOF, expire and defrag, but not for others.
why do you think it's half baked? maybe the "other" is, but not the rest...

maybe we can drop "other" and keep the IO / AOF? in theory users (or even the server) can deduct the IO and AOF from the total and conclude how much the rest takes, but i was concerned that there might be other things so i wanted to measure beforeSleep and serverCron so it'll be possible to find out that there's something else that we don't measure.

Reply:
So do we have a conclusion here?

Reply:
> why do you think it's half baked? maybe the "other" is, but not the rest...

@oranagra I don't think we're working back from end users and what problems they will be solving with this information. The original issue was about command latency. In the thread we identified that event loop time was an easy an accessible metric that basically gave this information. As EL time increased, latency would likely go up proportional to it.  So, `eventloop_duration` metrics seems useful. 

Command information also seems useful. End users have the detailed breakdown of all the commands, they can go and see what commands are taking a lot of time and make some decisions based off of that. We also already have this information being calculated, so it seems useful to have it aggregated. 

I'm not sure what the other metrics are really adding. If they see high write IO, what are users going to do about it. I want to keep as much code off the hot path as possible and I want Redis to be as simple as possible.

Reply:
@madolson you're right that some of these are not useful for users, but in some cases metrics in INFO are for us (developers) not users. i.e. if someone submits an issue, we ask for INFO output and try to figure out what's wrong in their deployment and how we can either improve redis, or advise them what's their problem (even if they can't make sense of these metrics).

i'm not saying i'm certain which way to go here, just saying i'd like to have more detail next time i'm looking at such issues and be able to know where the time is spent.
maybe we should add a section for `experimental` or `dev` / `debug` info fields?

p.s. i think that in anyway, our backwards compatibility contract for info fields is a little bit more flexible than commands, and we can drop / change some (ones we think are unlikely to be used by an application / script)

Reply:
I'll throw in my $0.02 and agree with @madolson.  In my experience, the one metric here that is useful is the event-loop cycle time.  Under heavy load, once Redis caps out on throughput, the event loop cycle time increases.  In extreme cases, with slow & pipelined commands being executed, I have observed scenarios where the event loop cycle time reaches several MINUTES in length.

It's important to consider this edge case.  How will this "instantaneous" metric function if the event loop takes 5 minutes to cycle?

IMO, 2 metrics would be useful:
* A rolling average of event loop cycle time over roughly a minute
* An "instantaneous" cycle time (over roughly a second)

And we need to understand how both of these will perform when the cycle time is insanely large (like 5 minutes).

Also, I agree that adding metrics just because we can is a dangerous strategy.  Our metrics are not clearly documented.  End-users don't know how to interpret some of the values.  Each metric increases the real-time requirements for INFO ALL.  And (probably most importantly) each metric adds a backwards compatibility burden.  An ill-conceived metric must be maintained for backward compatibility purposes.

@oranagra, I agree that maybe an "experimental" section might be a solution.  If so, it should be explicitly stated that items in that section will change without notice and will not be considered a backward-compatibility concern.  I'd suggest that:
* The experimental sections should not appear for INFO ALL
* The experimental sections should be clearly documented that they may change without notice
* We should consider adding this documentation to the live output...  something like:
```
# EventLoop (EXPERIMENTAL)
experimental: this section may change without notice
<metric>: <value>
...
```
* We could add a parameter on the INFO command like `INFO EXPERIMENTAL <section>`

Reply:
> It's important to consider this edge case. How will this "instantaneous" metric function if the event loop takes 5 minutes to cycle?

How about simply exposing an event loop counter?